WordPress/AbstractValidSlugSniff.php

```diff
@@ -0,0 +1,276 @@
+
+	/**
+	 * All valid tokens for the slug parameter.
+	 *
+	 * Set in `register()`.
+	 *
+	 * @var array<int|string, int|string>
+	 */
+	private $valid_tokens = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 3.0.0
+	 */
+	public function __construct() {
+		$this->target_functions = $this->get_target_functions();
+		$this->slug_type        = $this->get_slug_type();
+		$this->slug_type_plural = $this->get_slug_type_plural();
+		$this->valid_characters = $this->get_valid_characters();
+		$this->max_length       = $this->get_max_length();
+		$this->reserved_names   = $this->get_reserved_names();
+	}
+
+	/**
+	 * Retrieve function and parameter(s) pairs this sniff is looking for.
+	 *
+	 * The parameter or an array of parameters keyed by target function.
+	 * An array of names is supported to allow for functions for which the
+	 * parameter names have undergone name changes over time.
+	 *
+	 * @return array<string, string|array<string>> Function parameter(s) pairs.
+	 */
+	abstract protected function get_target_functions();
+
+	/**
+	 * Retrieve the slug type.
+	 *
+	 * @return string The slug type.
+	 */
+	abstract protected function get_slug_type();
+
+	/**
+	 * Retrieve the plural slug type.
+	 *
+	 * @return string The plural slug type.
+	 */
+	abstract protected function get_slug_type_plural();
+
+	/**
+	 * Retrieve the regex to validate the characters that can be used as
+	 * the slug.
+	 *
+	 * @return string Regular expression.
+	 */
+	abstract protected function get_valid_characters();
+
+	/**
+	 * Retrieve the max length of a slug.
+	 *
+	 * The length is limited by the SQL field.
+	 *
+	 * @return int The maximum length of a slug.
+	 */
+	abstract protected function get_max_length();
+
+	/**
+	 * Retrieve the reserved names which can not be used by themes and plugins.
+	 *
+	 * @return array<string, true> Key is reserved name, value irrelevant.
+	 */
+	abstract protected function get_reserved_names();
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @return array
+	 */
+	public function register() {
+		$this->valid_tokens = Tokens::$textStringTokens + Tokens::$heredocTokens + Tokens::$emptyTokens;
+		return parent::register();
+	}
+
+	/**
+	 * Process the parameter of a matched function.
+	 *
+	 * Errors on invalid names when reserved keywords are used,
+	 * the name is too long, or contains invalid characters.
+	 *
+	 * @param int    $stackPtr        The position of the current token in the stack.
+	 * @param string $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched
+	 *                                in lowercase.
+	 * @param array  $parameters      Array with information about the parameters.
+	 *
+	 * @return void
+	 */
+	public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) {
+		$slug_param = PassedParameters::getParameterFromStack(
+			$parameters,
+			1,
```

Comment:
> The parameter name must be provided by the concrete sniff. If that's the case, the offset is superseded unless the parameter name is incorrect. Am I understanding this correctly?

You are totally misunderstanding this.

In PHP, parameters can be passed _positionally_, i.e. `my_function($first, $second, $third)` or _by name_, i.e. `my_function($first, third: $value)`.

To get the correct parameter from a function call, the `PassedParameters::getParameterFromStack()` method needs to know **both** the parameter name(s) and the parameter position.

And as this is an abstract sniff, the sniff cannot rely on the "slug"-like parameter always being in the first position.

Reply:
I see! I thought this was checking against the name in the function declaration, rather than checking for named parameters. My mistake. If I had been more attentive, I would have noticed the line, 'First check for a named parameter.' As you may have noticed, I'm not very familiar with the WPCS or the PHPCS codebase. 😅 😇

Reply:
PHPCS has no access to the function declaration (unless the function is declared in the same file). For this type of code analysis, you only look at the function _call_.